From: Armbian Build Framework <build@armbian.com>
Subject: [PATCH] spl: sunxi: Fix SPL heap exhaustion during FIT loading and add eMMC settle delay

In SPL with CONFIG_SPL_MMC_TINY, malloc heap is limited (8KB), causing
malloc_cache_aligned to fail when allocating the 850KB FIT image buffer.
Route board_spl_fit_buffer_addr and spl_get_load_buffer directly to DRAM
at 0x41000000. Additionally, add a 250ms settle delay after mmc_init to allow
KLMAG2WEPD eMMC to complete internal power-on sequencing before reading data.

---
 common/spl/spl.c     |  2 +-
 common/spl/spl_fit.c | 16 +---------------
 common/spl/spl_mmc.c |  2 ++
 3 files changed, 4 insertions(+), 16 deletions(-)

--- a/common/spl/spl.c
+++ b/common/spl/spl.c
@@ -255,7 +255,7 @@
 
 __weak struct legacy_img_hdr *spl_get_load_buffer(ssize_t offset, size_t size)
 {
-	return map_sysmem(CONFIG_TEXT_BASE + offset, 0);
+	return (void *)(0x41000000 + offset);
 }
 
 void spl_set_header_raw_uboot(struct spl_image_info *spl_image)
--- a/common/spl/spl_fit.c
+++ b/common/spl/spl_fit.c
@@ -595,21 +595,7 @@
  */
 __weak void *board_spl_fit_buffer_addr(ulong fit_size, int sectors, int bl_len)
 {
-	void *buf;
-	size_t size = sectors * bl_len;
-
-	buf = malloc_cache_aligned(size);
-	if (!buf) {
-		pr_err("Could not get FIT buffer of %lu bytes\n", (ulong)size);
-
-		if (IS_ENABLED(CONFIG_SPL_SYS_MALLOC))
-			pr_err("\tcheck CONFIG_SPL_SYS_MALLOC_SIZE\n");
-		else
-			pr_err("\tcheck CONFIG_SPL_SYS_MALLOC_F_LEN\n");
-
-		buf = spl_get_load_buffer(0, size);
-	}
-	return buf;
+	return (void *)0x41000000;
 }
 
 /*
--- a/common/spl/spl_mmc.c
+++ b/common/spl/spl_mmc.c
@@ -395,6 +395,8 @@
 			return ret;
 		}
 	}
+	extern void udelay(unsigned long usec);
+	udelay(250000);
 
 	boot_mode = spl_mmc_boot_mode(mmc, bootdev->boot_device);
 	ret = -EINVAL;
